File: tests/xcalablemp/NOT-IMPLEMENTED/292/reduction_logical_and.f90
program main
  include 'xmp_lib.h'
  integer,parameter:: N=4
  integer result
  logical a(N), sa
!$xmp nodes p(*)
!$xmp template t(N)
!$xmp distribute t(cyclic) onto p
!$xmp align a(i) with t(i)

!$xmp loop (i) on t(i)
  do i=1, N
     a(i) = .FALSE.
  enddo

!$xmp loop (i) on t(i) reduction(.and.:sa)
  do i=1, N
     sa = a(i)
  end do

  if( sa .neqv. .FALSE.) then
     result = -1
  endif

!$xmp task on p(1)
  a(1) = .TRUE.
!$xmp end task

!$xmp loop (i) on t(i) reduction(.and.:sa)
  do i=1, N
     sa = a(i)
  end do

  if( sa .neqv. .FALSE.) then
     result = -1
  endif

!$xmp loop (i) on t(i)
  do i=1, N
     a(i) = .TRUE.
  enddo
  
!$xmp loop (i) on t(i) reduction(.and.:sa)
  do i=1, N
     sa = a(i)
  end do

  if( sa .neqv. .TRUE.) then
     result = -1
  endif
      
!$xmp reduction(+:result)
!$xmp task on p(1)
  if( result .eq. 0 ) then
     write(*,*) "PASS"
  else
     write(*,*) "ERROR"
     call exit(1)
  endif
!$xmp end task

end program main

File: tests/XMP/global-view/loop/C-Round/reduction_logical_and.c
#include <xmp.h>
#include <stdio.h>  
#include <stdlib.h> 
#define TRUE  1
#define FALSE 0
int N=4,i,a[N], sa=FALSE, result=0;
#pragma xmp nodes p(4)
#pragma xmp template t(0:N-1)
#pragma xmp distribute t(cyclic) onto p
#pragma xmp align a[i] with t(i)

int main(void)
{
#pragma xmp loop on t(i)
  for(i=0;i<N;i++)
    a[i] = FALSE;

#pragma xmp loop on t(i) reduction(&&:sa)
  for(i=0;i<N;i++)
    sa = a[i];    

  if(sa != FALSE)  result = -1;

#pragma xmp task on p(1)
  {
    a[0] = TRUE;
  }

#pragma xmp loop on t(i) reduction(&&:sa)
  for(i=0;i<N;i++)
    sa = a[i];

  if(sa != FALSE) result = -1;

#pragma xmp loop on t(i)
  for(i=0;i<N;i++)
    a[i] = TRUE;

#pragma xmp loop on t(i) reduction(&&:sa)
  for(i=0;i<N;i++)
    sa = a[i];

  if(sa != TRUE) result = -1;

#pragma xmp reduction(+:result)
#pragma xmp task on p(1)
  {
    if(result == 0){
      printf("PASS\n");
    }
    else{
      fprintf(stderr, "ERROR\n");
      exit(1);
    }
  }

  return 0;
}
